crates/router/src/workflows/revenue_recovery.rs

```diff
@@ -485,8 +546,334 @@ impl From<InternalDeciderRequest> for external_grpc_client::DeciderRequest {
+            Ok(best_token_and_time)
+        }
+    }
+}
+
+#[cfg(feature = "v2")]
+pub async fn calculate_smart_retry_time(
+    state: &SessionState,
+    merchant_id: &id_type::MerchantId,
+    payment_attempt: &PaymentAttemptResponse,
+    payment_intent: &PaymentsIntentResponse,
+    token_with_retry_info: &PaymentProcessorTokenWithRetryInfo,
+) -> Result<Option<time::PrimitiveDateTime>, errors::ProcessTrackerError> {
+    let wait_hours = token_with_retry_info.retry_wait_time_hours;
+    let monthly_retry_remaining = token_with_retry_info.monthly_retry_remaining;
+    let current_time = time::OffsetDateTime::now_utc();
+    let future_time = current_time + time::Duration::hours(wait_hours);
+
+    let future_timestamp = Some(prost_types::Timestamp {
+        seconds: future_time.unix_timestamp(),
+        nanos: 0,
+    });
+
+    get_schedule_time_for_smart_retry(
+        state,
+        merchant_id,
+        payment_attempt,
+        payment_intent,
+        monthly_retry_remaining,
+        future_timestamp,
+    )
+    .await
+}
+
+#[cfg(feature = "v2")]
+async fn process_token_for_retry(
+    state: &SessionState,
+    merchant_id: &id_type::MerchantId,
+    token_with_retry_info: &PaymentProcessorTokenWithRetryInfo,
+    payment_intent: &PaymentsIntentResponse,
+    payment_attempt: &PaymentAttemptResponse,
+) -> Result<Option<ScheduledToken>, errors::ProcessTrackerError> {
+    let token_status = &token_with_retry_info.token_status;
+    let inserted_by_attempt_id = &token_status.inserted_by_attempt_id;
+
+    let skip = decide_retry_failure_action(state, payment_attempt).await?;
+
+    match skip {
+        true => {
+            logger::info!(
+                "Skipping decider call due to hard decline for attempt_id: {}",
+                inserted_by_attempt_id.get_string_repr()
+            );
+            Ok(None)
+        }
+        false => {
+            let schedule_time = calculate_smart_retry_time(
+                state,
+                merchant_id,
+                payment_attempt,
+                payment_intent,
+                token_with_retry_info,
+            )
+            .await?;
+            Ok(schedule_time.map(|schedule_time| ScheduledToken {
+                token_details: token_status.payment_processor_token_details.clone(),
+                schedule_time,
+            }))
+        }
+    }
+}
+
+#[cfg(feature = "v2")]
+#[allow(clippy::too_many_arguments)]
+pub async fn call_decider_for_payment_processor_tokens_select_closet_time(
+    state: &SessionState,
+    merchant_context: domain::MerchantContext,
+    merchant_id: &id_type::MerchantId,
+    req_state: &ReqState,
+    profile: &domain::Profile,
+    processor_tokens: &HashMap<String, PaymentProcessorTokenWithRetryInfo>,
+    payment_id: &id_type::GlobalPaymentId,
+    connector_customer_id: &str,
+) -> CustomResult<Option<ScheduledToken>, errors::ProcessTrackerError> {
+    let payment_intent_response = Box::pin(payments::payments_intent_core::<
```

Comment:
These api related things are already defined in revenue_recovery api's reuse them